### DISCUSSION TITLE
External C function evaluation problem

### DISCUSSION THREAD

**Discussion by luizmac** [OpenModelica/OpenModelica#10831] [Q&A]
Hello,

I am writing a block to call an C function to integrate an input. However if I put more than one input the output changes when the other inputs change their values. Even that the other inputs are not being used in the C function, How this is possible?  

The model is:

![image](https://github.com/OpenModelica/OpenModelica/assets/67111542/60cb41ec-2254-40b8-9ea5-64c36828e913)

The input which I want to integrate is the first step source which starts in 0. But when the second and third sources start the integration presents a small change.

![image](https://github.com/OpenModelica/OpenModelica/assets/67111542/2550531a-0c6a-4b01-8be2-0c2b23a22a2d)

My C code does not use the second and third sources. Here is it:

-----------------------------------------------------------------------

```c
#ifndef _INTEGRATOR_C_
#define _INTEGRATOR_C_

integrador(double in1, double in2, double in3) {

  double y;
  static yant=0;

  y=in1+yant;
  yant=y;

  return y;
}

#endif
```

----------------------------------------------------------------------------

Why this is happening? How to avoid it?

My models are:

-------------------------------------------------------------------------

```mo
model Integrador_C
  Modelica.Blocks.Sources.Step step(height = 1, startTime = 0) annotation(
    Placement(visible = true, transformation(origin = {-70, -30}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Microcontrolador microcontrolador annotation(
    Placement(visible = true, transformation(origin = {-30, -56}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.Step step1(height = 1, startTime = 0.2) annotation(
    Placement(visible = true, transformation(origin = {-70, -64}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.Step step2(height = 1, startTime = 0.3) annotation(
    Placement(visible = true, transformation(origin = {-70, -98}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(step.y, microcontrolador.in1) annotation(
    Line(points = {{-58, -30}, {-54, -30}, {-54, -51}, {-41, -51}}, color = {0, 0, 127}));
  connect(step1.y, microcontrolador.in2) annotation(
    Line(points = {{-58, -64}, {-56, -64}, {-56, -56}, {-41, -56}}, color = {0, 0, 127}));
  connect(step2.y, microcontrolador.in3) annotation(
    Line(points = {{-58, -98}, {-41, -98}, {-41, -60}}, color = {0, 0, 127}));
  annotation(
    uses(Modelica(version = "4.0.0")),
    Diagram(coordinateSystem(extent = {{-100, 0}, {0, -120}})),
    version = "");
end Integrador_C;
```
----------------------------------------------------------------------------

```mo
model Microcontrolador

extends Modelica.Blocks.Icons.Block;
  Modelica.Blocks.Interfaces.RealInput in1 "comment" annotation (Placement(
        visible = true, transformation(origin = {0, 40}, extent = {{-120, -10}, {-100, 10}}, rotation = 0), iconTransformation(origin = {0, 50}, extent = {{-120, -10}, {-100, 10}}, rotation = 0)));
  Modelica.Blocks.Interfaces.RealInput in2 annotation(
    Placement(visible = true, transformation(origin = {0, 0}, extent = {{-120, -10}, {-100, 10}}, rotation = 0), iconTransformation(origin = {0, 0}, extent = {{-120, -10}, {-100, 10}}, rotation = 0)));
  Modelica.Blocks.Interfaces.RealInput in3 annotation(
    Placement(visible = true, transformation(origin = {0, -40}, extent = {{-120, -10}, {-100, 10}}, rotation = 0), iconTransformation(origin = {0, -42}, extent = {{-120, -10}, {-100, 10}}, rotation = 0)));
  Modelica.Blocks.Interfaces.RealOutput out1 "comment" annotation (Placement(
        visible = true, transformation(origin = {0, 20}, extent = {{100, -10}, {120, 10}}, rotation = 0), iconTransformation(origin = {0, 30}, extent = {{100, -10}, {120, 10}}, rotation = 0)));

function integrador "Modelica wrapper for an embedded C controller"
  input Real in1;
  input Real in2;
  input Real in3;
  output Real out1;
  external "C" annotation (Include="#include \"/mnt/Dados/Estudos/Modelica/Maccari_Lib/integrador.c\"");
end integrador;

algorithm // Algoritmo do bloco

  out1 := integrador(in1,in2,in3);  // Chamar aqui sempre todas as entradas


end Microcontrolador;
```
-----------------------------------------------------

**Comment by casella** [comment]
@luizmac you can't integrate an input inside an external function. Integration should be performed at the equation level by writing something as `der(x) = ...`

You can of course put a C function call on the right hand side, e.g. `der(x) = my_function(x, y, time)`, but then the function will compute the state _derivative_. The integration will be done by the Modelica tool.

**Reply by luizmac** [reply]
Hi,

I not trying to integrate using equation statement. I trying to put a discrete integration (Euler) in the external C code.

 I want to increase the value at every step of the simulation. I expect a perfect ramp. It is a simple accumulator. However when other inputs are inserted there is an increment not expected.

As you can see my c code has only one input the other two shouldn't insert anyhing in the ramp.

I trying to create a software in the loop simulation and my controller has an integrator which need to be incremented at every step. The discontinuity on the ramp shouldn't be present. 

It seems that when the other sources are inserted the external function is evaluated more than one time for the same instant.

Could it be a bug?

Best Regards

**Reply by casella** [reply]
> I not trying to integrate using equation statement. I trying to put a discrete integration (Euler) in the external C code.

That's exactly my point. You can't do that in Modelica. Time integration is not something you can write _explicitly_ in Modelica (i.e. by writing in Modelica an integration method). Unless you have clocked systems or discrete variables with periodic sampling time, in which case you are implicitly assuming some kind of fixed-time-step method will be used. But that requires event-handling and is usually not very efficient.

> I want to increase the value at every step of the simulation. I expect a perfect ramp.

But why should it be a ramp? Who guarantees that your function is called at regular time intervals? How is the time step (which could be variable) communicated to the C block? Who guarantees that your function is not called to perform a long step and then this step is rejected by the ODE solver because it fails the error control checks and retried with a shorter time step? How could you undo the accumulation in that case?

> It is a simple accumulator. However when other inputs are inserted there is an increment not expected.

That's precisely because the semantics of a Modelica model is not of something that is regularly executed at fixed time steps. It's a hybrid system of possibly stiff, nonlinear ODEs that are expected to be handled by some ODE or DAE solver _which is outside of the model_, on a higher level of abstraction.

Hope this helps :)

The real question is, why are you doing this? What are your requirements? Can you elaborate on that? Maybe there is some simpler way to do it in Modelica without writing external C code.

**Reply by luizmac** [reply]
Hi,

Now I got it. That issue in the ramp is probably because the external function is being called two times for the same time. Because I have used a fixed time step, I didn't expect that this behavior could be possible.  

Now I tested the simulation using a sample before the function call

algorithm
when sample(0, Ts) then 
  out1 := integrador(in1,in2,in3);  
end when;

Using the same fixed time step of the simulation my ramp grows ok without any issues.

I am trying to use a C function to emulate my C code used in a microprocessor. I have control libraries in C that I want to use in my simulation.  It is a Software-in-the-loop simulation. 

This type of simulation is easy handled with other simulators as PSIM or Simulink. I trying to transport that frame work to openmodelica.

The simulation has a counter which needs to be increased at every time step in order to create an internal triangular wave. This is the reason to get an accumulator in the external C function.  

Thanks for the help!

**Comment by AnHeuermann** [comment]
You forgot to add the return type of your C function `integrador`. Your Modelica function `integrador` suggest it should be `double`, but the default return type in C is `int`.

```diff
#ifndef _INTEGRATOR_C_
#define _INTEGRATOR_C_

- integrador(double in1, double in2, double in3) {
+ double integrador(double in1, double in2, double in3) {

  double y;
  static yant=0;

  y=in1+yant;
  yant=y;

  return y;
}

#endif
```

Besides that I could not reproduce your issue.

**Comment by casella** [comment]
@luizmac, if you want to simulate your digital controller inside a Modelica model that also contains a continuous time physical process model, you need to explicitly model the ADC/DAC interfaces, with their fixed sampling time. You can do that either via discrete variables and when sample() equations, or via clocked systems.

**Reply by luizmac** [reply]
Now I am using the sample structure. But a new problem arise. 

Is it possible to return more than one value from the external C function? My project depend on having multiple outputs from an external c function.

**Reply by casella** [reply]
Sure, see the Modelica Language Specification, [Section 12.9](https://specification.modelica.org/maint/3.6/functions.html#external-function-interface).
